src/Orleans.Runtime/Activation/IGrainContextActivator.cs

```diff
@@ -345,18 +347,67 @@ public bool TryGetConfigurator(GrainType grainType, GrainProperties properties,
                 throw new InvalidOperationException(
                     $"Wrong signature of callback method {callbackMethodName} " +
                     $"specified in MayInterleave attribute for grain class {grainType.FullName}. \n" +
-                    $"Expected: public static bool {callbackMethodName}(IInvokable req)");
+                    $"Expected: public bool {callbackMethodName}(IInvokable req)");
+            }
+
+            if (method.IsStatic)
+            {
+                return new MayInterleaveStaticPredicate(method.CreateDelegate<Func<IInvokable, bool>>());
             }
 
-            return method.CreateDelegate<Func<IInvokable, bool>>();
+            var functionType = Expression.GetFuncType(grainType, typeof(IInvokable), typeof(bool));
+            var functionDelegate = method.CreateDelegate(functionType);
+            var predicateType = typeof(MayInterleaveInstancedPredicate<>).MakeGenericType(grainType);
+
+            return Activator.CreateInstance(predicateType, functionDelegate) as IMayInterleavePredicate;
+        }
+    }
+
+    internal interface IMayInterleavePredicate
+    {
+        bool Invoke(object instance, IInvokable bodyObject);
+    }
+
+    internal class ReentrantPredicate : IMayInterleavePredicate
+    {
+        private ReentrantPredicate()
+        {
+        }
+
+        public static ReentrantPredicate Instance { get; } = new();
+
+        public bool Invoke(object _, IInvokable bodyObject) => true;
+    }
+
+    internal class MayInterleaveStaticPredicate : IMayInterleavePredicate
+    {
+        private readonly Func<IInvokable, bool> _mayInterleavePredicate;
+
+        public MayInterleaveStaticPredicate(Func<IInvokable, bool> mayInterleavePredicate)
+        {
+            _mayInterleavePredicate = mayInterleavePredicate;
         }
+
+        public bool Invoke(object _, IInvokable bodyObject) => _mayInterleavePredicate(bodyObject);
+    }
+
+    internal class MayInterleaveInstancedPredicate<T> : IMayInterleavePredicate where T : class
+    {
+        private readonly Func<T, IInvokable, bool> _mayInterleavePredicate;
```

Comment:
I believe this will work just the same with `object` in place of `T`, allowing us to avoid `MakeGenericType`. No need for you to make a change here: I will either merge this as-is or push a commit myself.

Reply:
Pushed

Reply:
Weird, I think I was getting exception when trying to create Func<object, IInvokable, bool> delegate, which is why I used generics

Reply:
I forgot to actually use instanced version of MayInterleave method in MayInterleaveInstancedPredicateGrain, which is why tests are still going through, my bad

Reply:
I pushed fixed version of grain that is used for tests, tests should fail now. This needs to be reverted or reimplemented some other way